### PR TITLE
Add govuk notify

### PR DESCRIPTION
**Pull request by kmahern**
#### What
Send devise emails from govuk notify.

#### Ticket

https://dsdmoj.atlassian.net/browse/CACP-122

#### Why
In order to notify users of password changes, email changes and password resets.

#### How

--------

### CODE REVIEW

File: app/mailers/devise_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+class DeviseMailer < GovukNotifyRails::Mailer
+  include Devise::Controllers::UrlHelpers
+  include ActionView::Helpers::DateHelper
+
+  def reset_password_instructions(user, token, _opts = {})
+    set_template('7314475f-62b0-41bf-8aff-768427cba456')
```
Comment (jsugarman) [inline app/mailers/devise_mailer.rb]: extract template ids to method, constant or mixin, can pair on this:

```
  # example only
  GOVUK_NOTIFY_TEMPLATES = {
    reset_password_instructions: 'that-long-id',
    emailed_changed: 'that-long-id',
   }
```
and use
```
set_template(GOVUK_NOTIFY_TEMPLATES[:reset_password_instructions))

```

File: config/environments/production.rb [hunk 2]
```diff
@@ -103,4 +103,9 @@
   # config.active_record.database_selector = { delay: 2.seconds }
   # config.active_record.database_resolver = ActiveRecord::Middleware::DatabaseSelector::Resolver
   # config.active_record.database_resolver_context = ActiveRecord::Middleware::DatabaseSelector::Resolver::Session
+
+  # Don't care if the mailer can't send.
+  config.action_mailer.raise_delivery_errors = false
```
Comment (jsugarman) [inline config/environments/production.rb]: we probably do want to raise errors?

File: app/mailers/devise_mailer.rb (hunk 1 above)
Comment (kmahern) [reply app/mailers/devise_mailer.rb]: 👍

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -14,7 +16,7 @@ def create
     authorize!(:create, @user)
 
     if @user.save
-      @user.send_reset_password_instructions
+      deliver_reset_password_instructions(@user)
```
Comment (jsugarman) [inline app/controllers/users_controller.rb]: it would be good to have this method called form the user object

File: app/mailers/devise/mailer.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      mail(to: user.email)
+    end
+
+    def email_changed(user, _opts = {})
+      set_template(GOVUK_NOTIFY_TEMPLATES[:email_changed])
+      set_personalisation(
+        user_full_name: user.name,
+        user_new_email: user.email
+      )
+      mail(to: user.email_before_last_save)
```
Comment (jsugarman) [inline app/mailers/devise/mailer.rb]: **Not a blocker** according to devise `email_was`, the default, should work 

https://github.com/heartcombo/devise/blob/a17abad57a5d97594701cee7eac072170f739313/lib/devise/models/database_authenticatable.rb#L176-L186

File: config/environments/test.rb [hunk 5]
```diff
@@ -21,6 +21,9 @@
     'Cache-Control' => "public, max-age=#{1.hour.to_i}"
   }
 
+  # Don't care if the mailer can't send.
+  config.action_mailer.raise_delivery_errors = false
```
Comment (jsugarman) [inline config/environments/test.rb]: should match whatever we use in production

**Comment by jsugarman** [comment]
could also add https://github.com/ministryofjustice/govuk_notify_rails#mailer-previewing

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (kmahern) [reply app/controllers/users_controller.rb]: I have removed password_helpers and this method is now called from the user object.

**Comment by kmahern** [comment]
Backgrounding of email generation is in background-GovukNotify branch.

**Pull request merged**